katdal/visdatav4.py

```diff
@@ -76,6 +76,57 @@ def _calc_azel(cache, name, ant):
     return sensor_data
 
 
+def _calc_delay(cache, name, inp):
+    """Extract virtual applied delay/phase sensors from raw CBF sensors."""
+    # Obtain the relevant CBF attributes from the cache
+    instrument = cache.get('Correlator/instrument')[0]
+    sync_time = cache.get('Correlator/sync_time')[0]
+    scale_factor_timestamp = cache.get('Correlator/scale_factor_timestamp')[0]
+    # Don't extract the sensor since the real timestamps are hidden in the values
+    getter = cache.get(f'{instrument}_antenna_channelised_voltage_{inp}_delay',
+                       extract=False)
+    sensor_data = getter.get()
+    # Unwrap and unpack the five components of the CBF sensor
+    values = [ComparableArrayWrapper.unwrap(v) for v in sensor_data.value]
+    adc_sample_counts, delays, delay_rates, phases, phase_rates = zip(*values)
+    # Convert the ADC sample count of each delay update to proper Unix timestamp
+    times = sync_time + np.array(adc_sample_counts) / scale_factor_timestamp
+    # Use actual delay/phase rate used by F-engine to interpolate between updates.
+    # Insert interpolation endpoint just before next update for strict monotonicity.
+    next_times = np.r_[times[1:], 2 * times[-1] - times[-2]] - 1e-6
+    next_delays = delays + delay_rates * (next_times - times)
+    next_phases = phases + phase_rates * (next_times - times)
+    times = np.c_[times, next_times].ravel()
+    delays = np.c_[delays, next_delays].ravel()
+    phases = np.c_[phases, next_phases].ravel()
+    delay_data = SimpleSensorGetter(name, times, delays)
+    phase_data = SimpleSensorGetter(name, times, phases)
+    cache[name.replace('applied_phase', 'applied_delay')] = delay_data
+    cache[name.replace('applied_delay', 'applied_phase')] = phase_data
+    return delay_data if name.endswith('delay') else phase_data
+
+
+def _calc_gain(cache, name, inp):
+    """Extract virtual applied F-engine gain sensors from raw CBF sensors."""
+    instrument = cache.get('Correlator/instrument')[0]
+    # The real/imag parts are cast to int16 in the F-engine but the CBF sensor
+    # seems to report back the CAM request, so round them here to be more accurate
+    sensor_data = cache.get(f'{instrument}_antenna_channelised_voltage_{inp}_eq',
+                            transform=lambda g: np.array(g, dtype=np.complex64).round())
```

Comment:
- I don't know whether the transform applies to individual samples or to the whole timeseries, but could using `np.array` cause users to get back 0d arrays in place of scalars?
- While extremely unlikely, converting to complex64 then rounding could give a different result to directly rounding.

Reply:
- The transform applies to the individual values of a categorical sensor (had to look that up!). Contrary to popular (my) belief, it doesn't even apply to numerical sensors... However, the values in `*_eq` are lists of `n_chans` complex values, so they are (a) categorical and (b) not scalar. So the answer is no.
- This ordering happened because I first had only the complex64 cast and then later considered the rounding too. The sensor values are definitely low-resolution (i.e. 7-8 decimal places), so I don't think the cast will change anything. The real question is what CBF does inside CMC + corr2... Do they work with complex128 or complex64 before the cast to int16? Since you are now an insider, maybe you can tell me 😄

Reply:
> Since you are now an insider, maybe you can tell me

I'm only working on the next-gen correlator, so no insights into the current one. The case I'm thinking of is:
```python
> np.array(1.49999999+0j, np.complex64).round()
(2+0j)
> np.array(1.49999999+0j).round()
(1+0j)
```
I'm not sure it's worth worrying about though. If it becomes a problem we should fix the CBF sensors to report what is used.

Reply:
The real culprit is katcorelib (i.e. CAM or me :-)) that calls `?gain` with float real/imags while the ICD actually calls for integer real/imags. This is not helped by CBF generously accepting our invalid data and rounding it for us...

The upside is that we know what we would have liked it to be, and could correct for the quantisation effects (although that would be of little use in practice).